---
layout: lesson
root: .
permalink: index.html  # Is the only page that don't follow the partner /:path/index.html
---
**Welcome to R!** Working with a programming language (especially if it’s your
first time) often feels intimidating, but the rewards outweigh any frustrations.
An important secret of coding is that even experienced programmers find it
difficult and frustrating at times – so if even the best feel that way, why let
intimidation stop you? Given time and practice* you will soon find it easier
and easier to accomplish what you want.

**The workshop is organised by [the Open Science Community Saudi Arabia](https://osc-ksa.com/ar-index.html). It’ll be delivered by instructors and facilitators from the Carpentries in MENA.**

Why learn to code? Bioinformatics – like biology – is messy. Different
organisms, different systems, different conditions, all behave differently.
Experiments at the bench require a variety of approaches – from tested protocols
to trial-and-error. Bioinformatics is also an experimental science, otherwise we
could use the same software and same parameters for every genome assembly.
Learning to code opens up the full possibilities of computing, especially given
that most bioinformatics tools exist only at the command line. Think of it this
way: if you could only do molecular biology using a kit, you could probably
accomplish a fair amount. However, if you don’t understand the biochemistry of
the kit, how would you troubleshoot? How would you do experiments for which
there are no kits?

R is one of the most widely-used and powerful programming languages in
bioinformatics. R especially shines where a variety of statistical tools are
required (e.g. RNA-Seq, population genomics, etc.) and in the generation of
publication-quality graphs and figures. Rather than get into an R vs. Python
debate (both are useful), keep in mind that many of the concepts you will learn
apply to Python and other programming languages.

Finally, we won’t lie; R is not the easiest-to-learn programming language ever
created. So, don’t get discouraged! The truth is that even with the modest
amount of R we will cover today, you can start using some sophisticated R
software packages, and have a general sense of how to interpret an R script.
Get through these lessons, and you are on your way to being an accomplished R
user!

\* We very intentionally used the word practice. One of the other “secrets” of
programming is that you can only learn so much by reading about it. Do the
exercises in class, re-do them on your own, and then work on your own problems.


> ## Before you start
> - **Before the training:** Please make sure you have completed the [Setup](https://carpentries-incubator.github.io/open-science-with-r/setup.html) instructions.   
> - There are two options to install necessary softwares and packages:
>   * (Option 1) The favored one is to sign up to [RStudio Cloud](https://rstudio.cloud/).
>   * (Option 2) Alternatively, you can install everything manually. Follow the Setup instructions to do so.
>
> - Please read the workshop [Code of Conduct](https://github.com/Open-Science-Community-Saudi-Arabia/Open-Science-Community-in-Saudi/blob/main/CODE_OF_CONDUCT.md) to make sure this workshop stays welcoming for everybody.   
>
> - **Experimenter's Mindset**: We define the "Experimenter's mindset" as an
>    approach to bioinformatics that treats it like any other experiment. There
>    are probably a variety of metaphors we could employ (data are our
>    reagents, scripts are our protocols, etc.), but the most important idea of
>    the mindset is to remind you that as a researcher, you need to employ all
>    of your training in the bench or field to working with analyses. Evaluate
>    results critically, and don't expect that things will always work the first
>    time, or that they will always work in the same way.
{: .prereq}
